[EXPERIMENTAL] Integration Test on Swarm (completes in about 5 minutes, with 10 Swarm nodes)

AkihiroSuda opened the pull request:
## What I did
This PR adds `contrib/integration-cli-on-swarm`,
which enables IT to be running in parallel, using Swarm-mode and [Funker](https://github.com/bfirsh/funker).

This PR replaces my previous PR based on GNU parallel (hence it did not scale): https://github.com/docker/docker/pull/26684

Please refer to `contrib/integration-cli-on-swarm/README.md` for further information.

The test takes almost 5 or 6 minutes, with 10 [n1-standard-4](https://cloud.google.com/compute/docs/machine-types) (4 vCPUs + 15GB RAM per node) GCE instances, as of Dec 29, 2016.
(The revision used for this experimental result is https://github.com/AkihiroSuda/docker/commits/integration-cli-on-swarm.20161229-1 )
```console
$ ./contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh --push-worker-image example.gcr.io/foo/bar --replicas 30 --shuffle
(NOTE: CLI flags has changed from this 2016/12/29 revision) 
2016/12/29 08:32:15 Loaded 1618 tests (30 chunks)
2016/12/29 08:32:15 Executing 30 chunks in parallel, against "integration-cli-worker"
2016/12/29 08:32:15 Executing chunk 0 (contains 54 test filters)
..
2016/12/29 08:34:34 Finished chunk 28 [1/30] with 54 test filters in 2m19.098068003s, code=0.
2016/12/29 08:34:38 Finished chunk 12 [2/30] with 54 test filters in 2m23.088569511s, code=0.
2016/12/29 08:34:48 Finished chunk 10 [3/30] with 54 test filters in 2m33.880679079s, code=0.
2016/12/29 08:34:54 Finished chunk 20 [4/30] with 54 test filters in 2m39.973747028s, code=0.
2016/12/29 08:35:11 Finished chunk 18 [5/30] with 54 test filters in 2m56.28384361s, code=0.
2016/12/29 08:35:11 Finished chunk 29 [6/30] with 52 test filters in 2m56.54047088s, code=0.
2016/12/29 08:35:15 Finished chunk 1 [7/30] with 54 test filters in 3m0.285044426s, code=0.
2016/12/29 08:35:22 Finished chunk 6 [8/30] with 54 test filters in 3m7.211775338s, code=0.
2016/12/29 08:35:24 Finished chunk 25 [9/30] with 54 test filters in 3m9.938413009s, code=0.
2016/12/29 08:35:30 Finished chunk 27 [10/30] with 54 test filters in 3m15.219834368s, code=0.
2016/12/29 08:35:36 Finished chunk 9 [11/30] with 54 test filters in 3m21.615434162s, code=0.
2016/12/29 08:35:41 Finished chunk 13 [12/30] with 54 test filters in 3m26.576907401s, code=0.
2016/12/29 08:35:45 Finished chunk 17 [13/30] with 54 test filters in 3m30.290752537s, code=0.
2016/12/29 08:35:53 Finished chunk 2 [14/30] with 54 test filters in 3m38.148423321s, code=0.
2016/12/29 08:35:55 Finished chunk 24 [15/30] with 54 test filters in 3m40.09669137s, code=0.
2016/12/29 08:35:57 Finished chunk 8 [16/30] with 54 test filters in 3m42.299945108s, code=0.
2016/12/29 08:35:57 Finished chunk 22 [17/30] with 54 test filters in 3m42.946558809s, code=0.
2016/12/29 08:35:59 Finished chunk 23 [18/30] with 54 test filters in 3m44.232557165s, code=0.
2016/12/29 08:36:02 Finished chunk 3 [19/30] with 54 test filters in 3m47.112051358s, code=0.
2016/12/29 08:36:11 Finished chunk 15 [20/30] with 54 test filters in 3m56.340656645s, code=0.
2016/12/29 08:36:11 Finished chunk 11 [21/30] with 54 test filters in 3m56.882401231s, code=0.
2016/12/29 08:36:22 Finished chunk 19 [22/30] with 54 test filters in 4m7.551093516s, code=0.
2016/12/29 08:36:23 Finished chunk 21 [23/30] with 54 test filters in 4m8.221093446s, code=0.
2016/12/29 08:36:25 Finished chunk 16 [24/30] with 54 test filters in 4m10.450451705s, code=0.
2016/12/29 08:36:27 Finished chunk 5 [25/30] with 54 test filters in 4m12.162272692s, code=0.
2016/12/29 08:36:28 Finished chunk 14 [26/30] with 54 test filters in 4m13.977801031s, code=0.
2016/12/29 08:36:29 Finished chunk 0 [27/30] with 54 test filters in 4m14.34086812s, code=0.
2016/12/29 08:36:49 Finished chunk 26 [28/30] with 54 test filters in 4m34.437085539s, code=0.
2016/12/29 08:37:14 Finished chunk 7 [29/30] with 54 test filters in 4m59.22902721s, code=0.
2016/12/29 08:37:20 Finished chunk 4 [30/30] with 54 test filters in 5m5.103469214s, code=0.
2016/12/29 08:37:20 Executed 30 chunks in 5m5.104379119s. PASS: 30, FAIL: 0.
```

Since the makespan is significantly shorter than the current sequential CI, developers no longer need to wait for hours to catch failures.
Also, this PR will be useful for enabling `BUILDFLAGS=-race`, which is virtually impossible in the current CI because the race detector is slow. https://github.com/docker/docker/issues/26685
So, this will be a major step to race-free code.

As mentioned in https://github.com/docker/docker/pull/26684#issuecomment-253537257 , the intention of this work is not to replace the current sequential CI.

## How I did it

Using Swarm-mode and [Funker](https://github.com/bfirsh/funker).

Please refer to the attached image for the basic design.
![it on swarm](https://cloud.githubusercontent.com/assets/9248427/21541809/ca400c40-cdfc-11e6-8dc1-97d59bfd37d2.png)

Some implementation notes:

 * ~Currently, due to the limitation of Funker, the master container cannot fetch the log stream (e.g. `PASS: TestFooBar`..) from the workers, and hence we need to use `docker service logs WORKER` for fetching the log stream.~ (EDIT: now master fetch the logs from workers in batch)
 * Since swarm-mode service lacks the `--privileged` flag,  bind-mounted `docker.sock` is used for  running `docker run ... hack/make.sh ..` in the worker containers

## How to verify it

* Creates a swarm cluster (v1.13 ~with `--experimental` is required because the PR depends on `docker service logs`~)
* Creates a private registry (If you have multiple nodes)
* Run `make build-integration-cli-on-swarm` to create the images
* Run the `integration-cli-on-swarm` as follows:

```console
 $ ./contrib/integration-cli-on-swarm/integration-cli-on-swarm -push-worker-image example.gcr.io/foo/bar -replicas 30 -shuffle
```

As mentioned above, it took almost 5 or 6 minutes in my environment with 10 n1-standard-4 instances as of Dec 29 2016.

Notes:

  * `-push-worker-image` is not needed if you have only a single node
  * The larger the number of `replicas`, the higher parallelism, but also the higher overhead


### Travis CI result (just for demo)

The example of the actual output is available at https://travis-ci.org/AkihiroSuda/docker/builds/187416080

In this demo, only single free Travis instance  ( [up to 2 bursted vCPUs and 7.5GB RAM](https://docs.travis-ci.com/user/ci-environment/) ) is used.

The number of replicas is set to 5.

It took 32 minutes except the time spent in installing docker and building images.

```console
2016/12/29 09:36:35 Loaded 1618 tests (5 chunks)
...
2016/12/29 10:08:18 Executed 5 chunks in 31m43.682035889s. PASS: 2, FAIL: 3.
```

The failure seems because of flakiness in the test code and of ENOSPC.


## TODOs

Needs to be done before merging this PR:

 - [X] ~Vendor cpuguy83/check#4~ *EDIT*: Updated this PR, cpuguy83/check#4 is no longer needed
 - [x] Vendor bfirsh/funker-go#6
 - [x] Remove `.travis.yml`, which is just used for demo of this PR.
 - [x] Squash commits

Future work

- [X] Enables the master to fetch the worker logs (mentioned above)
- Use statistics rather than shuffling for mitigating the makespan nonuniformity
- Mitigates setup/teardown overhead
- [X] Enables `BUILDFLAGS=-race`
- [X] Enables `DOCKER_INCREMENTAL_BINARY`
- [X] Eliminate `os/exec` from `worker.go`
- [X] Rewrite bash part in Go

## Description for the changelog

N/A

## A picture of a cute animal (not mandatory but encouraged)

![image](https://cloud.githubusercontent.com/assets/9248427/21541840/ff366a8e-cdfc-11e6-8403-32be5416b009.png)

CC @vdemeester @bfirsh @justincormack @cpuguy83 @unclejack @dnephin @LK4D4 @mlaventure

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda committed:
[EXPERIMENTAL] Integration Test on Swarm

This commit adds contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh,
which enables IT to be running in parallel, using Swarm-mode and Funker.

Please refer to contrib/integration-cli-on-swarm/README.md

The test takes almost 5 to 6 minutes, with 10 n1-standard-4 GCE instances.

  $ ./contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh --push-worker-image example.gcr.io/foo/bar --replicas 30 --shuffle
  2016/12/29 08:32:15 Loaded 1618 tests (30 chunks)
  2016/12/29 08:32:15 Executing 30 chunks in parallel, against "integration-cli-worker"
  2016/12/29 08:32:15 Executing chunk 0 (contains 54 test filters)
  ..
  2016/12/29 08:34:34 Finished chunk 28 [1/30] with 54 test filters in 2m19.098068003s, code=0.
  2016/12/29 08:34:38 Finished chunk 12 [2/30] with 54 test filters in 2m23.088569511s, code=0.
  2016/12/29 08:34:48 Finished chunk 10 [3/30] with 54 test filters in 2m33.880679079s, code=0.
  2016/12/29 08:34:54 Finished chunk 20 [4/30] with 54 test filters in 2m39.973747028s, code=0.
  2016/12/29 08:35:11 Finished chunk 18 [5/30] with 54 test filters in 2m56.28384361s, code=0.
  2016/12/29 08:35:11 Finished chunk 29 [6/30] with 52 test filters in 2m56.54047088s, code=0.
  2016/12/29 08:35:15 Finished chunk 1 [7/30] with 54 test filters in 3m0.285044426s, code=0.
  2016/12/29 08:35:22 Finished chunk 6 [8/30] with 54 test filters in 3m7.211775338s, code=0.
  2016/12/29 08:35:24 Finished chunk 25 [9/30] with 54 test filters in 3m9.938413009s, code=0.
  2016/12/29 08:35:30 Finished chunk 27 [10/30] with 54 test filters in 3m15.219834368s, code=0.
  2016/12/29 08:35:36 Finished chunk 9 [11/30] with 54 test filters in 3m21.615434162s, code=0.
  2016/12/29 08:35:41 Finished chunk 13 [12/30] with 54 test filters in 3m26.576907401s, code=0.
  2016/12/29 08:35:45 Finished chunk 17 [13/30] with 54 test filters in 3m30.290752537s, code=0.
  2016/12/29 08:35:53 Finished chunk 2 [14/30] with 54 test filters in 3m38.148423321s, code=0.
  2016/12/29 08:35:55 Finished chunk 24 [15/30] with 54 test filters in 3m40.09669137s, code=0.
  2016/12/29 08:35:57 Finished chunk 8 [16/30] with 54 test filters in 3m42.299945108s, code=0.
  2016/12/29 08:35:57 Finished chunk 22 [17/30] with 54 test filters in 3m42.946558809s, code=0.
  2016/12/29 08:35:59 Finished chunk 23 [18/30] with 54 test filters in 3m44.232557165s, code=0.
  2016/12/29 08:36:02 Finished chunk 3 [19/30] with 54 test filters in 3m47.112051358s, code=0.
  2016/12/29 08:36:11 Finished chunk 15 [20/30] with 54 test filters in 3m56.340656645s, code=0.
  2016/12/29 08:36:11 Finished chunk 11 [21/30] with 54 test filters in 3m56.882401231s, code=0.
  2016/12/29 08:36:22 Finished chunk 19 [22/30] with 54 test filters in 4m7.551093516s, code=0.
  2016/12/29 08:36:23 Finished chunk 21 [23/30] with 54 test filters in 4m8.221093446s, code=0.
  2016/12/29 08:36:25 Finished chunk 16 [24/30] with 54 test filters in 4m10.450451705s, code=0.
  2016/12/29 08:36:27 Finished chunk 5 [25/30] with 54 test filters in 4m12.162272692s, code=0.
  2016/12/29 08:36:28 Finished chunk 14 [26/30] with 54 test filters in 4m13.977801031s, code=0.
  2016/12/29 08:36:29 Finished chunk 0 [27/30] with 54 test filters in 4m14.34086812s, code=0.
  2016/12/29 08:36:49 Finished chunk 26 [28/30] with 54 test filters in 4m34.437085539s, code=0.
  2016/12/29 08:37:14 Finished chunk 7 [29/30] with 54 test filters in 4m59.22902721s, code=0.
  2016/12/29 08:37:20 Finished chunk 4 [30/30] with 54 test filters in 5m5.103469214s, code=0.
  2016/12/29 08:37:20 Executed 30 chunks in 5m5.104379119s. PASS: 30, FAIL: 0.

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented on vendor/github.com/go-check/check/run.go:
```diff
@@ -23,28 +25,48 @@ func Suite(suite interface{}) interface{} {
+	newListFlag      = flag.Bool("check.list", false, "List the names of all tests that will be run")
+	newWorkFlag      = flag.Bool("check.work", false, "Display and do not remove the test working directory")
+	checkTimeout     = flag.String("check.timeout", "", "Panic if test runs longer than specified duration")
+	checkFiltersFile = flag.String("check.ffile", "", "A path to a LF-separated file that contains filter strings (OR-match)")
```
`-check.ffile` is not needed and removed in the PR for upstream: https://github.com/cpuguy83/check/pull/4

AkihiroSuda commented:
@vdemeester Can I hear your feedback :smile: ?

LK4D4 commented:
@vdemeester @dnephin what do you think? Thanks!

dnephin commented on contrib/integration-cli-on-swarm/agent/worker/worker.go:
```diff
@@ -0,0 +1,102 @@
+		"--entrypoint", "/bin/bash",
+		workerImage,
+		"hack/make.sh", "test-integration-cli",
+	)
```
It would be good to use the docker API here instead of exec.

AkihiroSuda replied:
Yes, I'll update

dnephin commented on contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh:
```diff
@@ -0,0 +1,235 @@
+#!/bin/bash
+# integration-cli-on-swarm.sh: run the integration tests in parall across a Swarm cluster
+# Please refer to README.md for the usage.
```
I wonder if we could find a way to write this driver in something other than bash.

Maybe a `docker-compose.yml` would help reduce the size of this script?

AkihiroSuda replied:
Maybe so. I'll try it.

dnephin reviewed (commented):
In general I think the design makes a lot of sense. I left some comment on more specific issues.

AkihiroSuda commented on Makefile:
```diff
@@ -142,6 +142,11 @@ test-docker-py: build ## run the docker-py tests
+test-integration-cli-on-swarm: ## run the integration tests in parall across a Swarm cluster
+# For ease of passing arguments (e.g. --replicas 3), the script is currently not started via Makefile
+	echo "Please use contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh"
+	contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh --help
```
Thank you for the feedback.
I'll try to provide functional `make` target (or just remove at the moment and revive in a future PR)

AkihiroSuda commented on contrib/integration-cli-on-swarm/agent/main.go:
```diff
@@ -0,0 +1,27 @@
+package main
+
```
I'm not sure it is possible, because `Dockerfile.worker` depends on `docker-dev:<current-branch>` so as to install the worker agent onto it.

AkihiroSuda commented:
Thank you @dnephin , I'll update this PR in short.
PTAL @vdemeester

LK4D4 commented:
Design makes sense for me as well. And results is exceptional, too.

AkihiroSuda commented:
Updated PR to use `docker stack`
Example output on a single free Travis node will be available at https://travis-ci.org/AkihiroSuda/docker/builds/197236352

AkihiroSuda commented:
Further updated to support DOCKER_INCREMENTAL_BINARY and BUILDFLAGS

vdemeester reviewed (approved):
This looks awesome ! and works 👼 👍 
As it is in `contrib/` for now it doesn't affect the current CI so we can iterate safely on it 😉.

LGTM 🦁

LK4D4 commented on .gitignore:
```diff
@@ -31,3 +31,5 @@ man/man1
 man/man5
 man/man8
 vendor/pkg/
+# generated by integration-cli-on-swarm
```
not sure where it is generated, but github complains that there is no last line break

AkihiroSuda replied:
Ah, sorry for confusing comment.
`docker-compose.yml` is generated by integration-cli-on-swarm, but `.gitignore` is still written by hand and I forgot to add line break.

I'll fix this (Actually no need to put yml there, I'll try to use /tmp/rand0m.yml instead)

AkihiroSuda commented on contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh:
```diff
@@ -0,0 +1,208 @@
+# integration-cli-on-swarm.sh: run the integration tests in parall across a Swarm cluster
+# Please refer to README.md for the usage.
+#
+# TODO(AkihiroSuda): rewrite in Go? (Maybe it would just result in increase of LOC?)
```
OK, let me rewrite in Go before merging 😍

AkihiroSuda commented on contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh:
```diff
@@ -0,0 +1,208 @@
+inspect_master_container_id() {
+    [ $# -eq 0 ]
+    # FIXME(AkihiroSuda): we should not rely on internal service naming convention
+    docker ps --all --quiet \
```
It is not sufficient to get the container ID (I need this ID to wait for its completion)

AkihiroSuda commented:
Rewrote bash part in Go

UX changes:

 * Now executable is located as `./contrib/integration-cli-on-swarm/integration-cli-on-swarm` (ELF binary), instead of ``./contrib/integration-cli-on-swarm/integration-cli-on-swarm.sh`
 * Now `--blah` is `-blah` (because using Go's standard `flag`)

Some implementation note:
 *  Stack is managed via `os/exec("docker stack ...")` rather than API (because there is no stack API :sweat_smile:)

vdemeester commented:
> Stack is managed via os/exec("docker stack ...") rather than API (because there is no stack API 😅)

Yeah 😅 That should come at some point 👼

AkihiroSuda commented:
Updated PR:

 - Rebased
 - Squashed commits
 - Removed unuseful `.travis.yml`
 -  Supported setting `DOCKER_EXPERIMENTAL` and `DOCKER_GRAPHDRIVER`. (Also mentioned them in `contrib/integration-cli-on-swarm/README.md`)
 

Because the `bundles` (and hence `$DEST`) directory is baked into image, I faced difficulty in making non-VFS driver work (e.g. `"'overlay2' is not supported over overlayfs"` error ) .

 So I added an new environment variable `DOCKER_INTEGRATION_DAEMON_DEST`, which is the equivalent of `$DEST` but can exist out of `bundles`. 


PTAL @vdemeester

AkihiroSuda commented:
With overlay2 graphdriver, now it completes in less than 4 minutes :smile:

```
2017/02/07 03:21:30 Loaded 1633 tests (40 chunks)
...
2017/02/07 03:25:12 Finished chunk 35 [40/40] with 41 test filters in 3m41.798660377s, code=0.
2017/02/07 03:25:12 Executed 40 chunks in 3m41.799095867s. PASS: 40, FAIL: 0.
```

vdemeester commented on contrib/integration-cli-on-swarm/agent/worker/worker.go:
```diff
@@ -0,0 +1,69 @@
+
+func validImageDigest(s string) bool {
+	// FIXME
+	return strings.Contains(s, "sha256:")
```
You could use `github.com/docker/distribution/reference` package here to validate the digest :angel:

thaJeztah replied:
Yes, I think that's a good idea :)

vdemeester reviewed (approved):
still LGTM 🐸
/cc @dnephin @icecrime @thaJeztah @aaronlehmann 

(as it's in contrib, I really feel it's mergeable quickly and we can iterate on it 👼)

AkihiroSuda commented:
done @vdemeester @thaJeztah

vdemeester reviewed (approved):
still LGTM 👼

tiborvass commented:
@AkihiroSuda I wonder if this shouldn't be in `hack/` instead of `contrib/` even if experimental.

AkihiroSuda commented:
done @tiborvass

thaJeztah reviewed (approved):
LGTM, lets go!

vdemeester commented:
🎉
